mmpose/models/losses/regression_loss.py

```diff
@@ -4,8 +4,107 @@
+        residual (bool): Option to add L1 loss and let the flow
+            learn the residual error distribution.
+    """
+
+    @staticmethod
+    def get_scale_net():
+        return nn.Sequential(
+            nn.Linear(2, 64), nn.LeakyReLU(), nn.Linear(64, 64),
+            nn.LeakyReLU(), nn.Linear(64, 2), nn.Tanh())
+
+    @staticmethod
+    def get_trans_net():
+        return nn.Sequential(
+            nn.Linear(2, 64), nn.LeakyReLU(), nn.Linear(64, 64),
+            nn.LeakyReLU(), nn.Linear(64, 2))
+
+    def __init__(self,
+                 use_target_weight=False,
+                 size_average=True,
+                 residual=True):
+        super(RLELoss, self).__init__()
+        self.size_average = size_average
+        self.use_target_weight = use_target_weight
+        self.residual = residual
+        self.amp = 1 / math.sqrt(2 * math.pi)
+
+        prior = distributions.MultivariateNormal(torch.zeros(2), torch.eye(2))
+        masks = torch.tensor([[0, 1], [1, 0]] * 3, dtype=torch.float32)
+        self.flow = RealNVP(self.get_scale_net, self.get_trans_net, masks,
+                            prior)
+
+    def _apply(self, fn):
+        self.flow.prior.loc = fn(self.flow.prior.loc)
+        self.flow.prior.scale_tril = fn(self.flow.prior.scale_tril)
+        self.flow.prior._unbroadcasted_scale_tril =\
+            fn(self.flow.prior._unbroadcasted_scale_tril)
+        self.flow.prior.covariance_matrix = \
+            fn(self.flow.prior.covariance_matrix)
+        self.flow.prior.precision_matrix = \
+            fn(self.flow.prior.precision_matrix)
+        self.flow.mask = fn(self.flow.mask)
+        self.flow.s._apply(fn)
+        self.flow.t._apply(fn)
+
+    def forward(self, output, target, target_weight=None):
+        """Forward function.
+
+        Note:
+            - batch_size: N
+            - num_keypoints: K
+            - dimension of keypoints: D (D=2 or D=3)
+
+        Args:
+            output (torch.Tensor[N, K, D*2]): Output regression,
+                    including coords and sigmas.
+            target (torch.Tensor[N, K, D]): Target regression.
+            target_weight (torch.Tensor[N, K, D]):
+                Weights across different joint types.
+        """
+        sigma = output[:, :, 2:].sigmoid()
+        coord = output[:, :, :2]
+
+        error = (coord - target) / sigma
+        # (B, K, 2)
+        log_phi = self.flow.log_prob(error.reshape(-1, 2))
+        log_phi = log_phi.reshape(target.shape[0], target.shape[1], 1)
+        log_sigma = torch.log(sigma).reshape(target.shape[0], target.shape[1],
+                                             2)
+        nf_loss = log_sigma - log_phi
+
+        if self.residual:
+            loss_sigma = torch.log(sigma / self.amp)
+            loss_l1 = torch.abs(target - coord)
+            loss_l1 /= (math.sqrt(2) * sigma + 1e-9)
+            loss = nf_loss + loss_l1 + loss_sigma
+        else:
+            loss = nf_loss
```

Comment:
I need to re-read the paper and original project for these questions. Is it ok to reply after some days?

Reply:
No problem, take your time, please. Thanks for your help!

Reply:
Sorry for the late reply, please refer to the attached file since in-line formula is not supported by Github.
[Reply.md](https://github.com/open-mmlab/mmpose/files/8424900/Reply.md)
[Reply.pdf](https://github.com/open-mmlab/mmpose/files/8424901/Reply.pdf)

Reply:
I simplified the implementation of log(Q) in commit [facb0d0](https://github.com/open-mmlab/mmpose/pull/1259/commits/facb0d01388b3216668ba3edb4bd3346c50c0bc6). 
Experiments are going on to check the influence.

Reply:
> I simplified the implementation of log(Q) in commit [facb0d0](https://github.com/open-mmlab/mmpose/pull/1259/commits/facb0d01388b3216668ba3edb4bd3346c50c0bc6). Experiments are going on to check the influence.

The experimental result is 86.2, 0.4 lower than the original implementation··· I am confused

Reply:
Sorry for the late reply.
First of all, the forward process of flow is:
$$
\bar\mu=\phi(z)\\
\mu = \bar\mu*\hat\sigma+\hat\mu
$$
where $\hat\mu$ and $\mu$ is the predicted and gt joints, $\hat\sigma$ is the predicted variance. 

1. The `error` in my code is $\bar\mu$ in the paper, which is:
   $$
   \bar\mu=f^{-1}(\mu)=\frac{\hat\mu-\mu}{\hat\sigma}
   $$
   We know that when  $y=f(x)$, we have:
   $$
   logP_f(y)=logP(x)+log|det(J(f^{-1}))|
   $$
   Thus, the loss without residual is:
   $$
   -logP(\mu)=-logP_\phi(\bar\mu)+log\ \hat\sigma
   $$
   where `self.flow.log_prob(error)` is $-logP(\bar\mu)$

2. For the ***double log(sigma)***, I think the author may mis-implemented the loss from that in the paper:
   $$
   -logP(\mu)=-logP(\bar\mu)+log\ \hat\sigma\\
   log_\phi P(\bar\mu) = logQ(\bar\mu)+log_\phi G(\bar\mu) + log s
   $$
   to:
   $$
   -logP(\mu)=-logP(\bar\mu)-logQ(\mu)+log\ \hat\sigma\\
   log_\phi P(\bar\mu) = log_\phi G(\bar\mu) + log s
   $$
   In the first formula, the variance for the distribution of $\bar\mu$ should be constant not $\hat\sigma$ since it was already re-scaled.

3. Yes, it's ignored, you can refer to the appendix of the paper

   > The term s barely affects the results and can be removed to save computation resources. Therefore, in our implementation, we drop the term log s for simplicity.

4. If I am right in the ***double log(sigma)*** question, I think it's unnecessary for this constant factor anymore.

Reply:
I simplified the implementation of log(Q) in commit [facb0d0](https://github.com/open-mmlab/mmpose/pull/1259/commits/facb0d01388b3216668ba3edb4bd3346c50c0bc6).

Reply:
@ly015 Sorry to bother, just wondering if you've seen my reply since I didn't @ you

Reply:
Yes. Sorry for the late reply. I will review the latest updates this week. Thank you very much.

Reply:
Ok! Thank you for your reply and review :wink: !